cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/NamespaceInitializerService.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.master.environment.k8s;
+
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import io.cdap.cdap.common.namespace.NamespaceQueryAdmin;
+import io.cdap.cdap.master.spi.namespace.NamespaceDetail;
+import io.cdap.cdap.master.spi.namespace.NamespaceListener;
+import org.apache.twill.api.TwillRunnerService;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A service for initializing the {@link TwillRunnerService} with a list of namespaces.
+ */
+public class NamespaceInitializerService extends AbstractExecutionThreadService {
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final TwillRunnerService twillRunnerService;
+  private List<NamespaceDetail> namespaceDetails;
+
+  NamespaceInitializerService(NamespaceQueryAdmin namespaceQueryAdmin, TwillRunnerService twillRunnerService) {
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.twillRunnerService = twillRunnerService;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    namespaceDetails = namespaceQueryAdmin.list().stream()
```

Comment:
I don't think we should do this on start up, but on run(). Generally we want to avoid doing things on startup that can take a long time or fail. It also allows us to skip this call entirely if the runner service is not a NamespaceListener

Reply:
fixed